iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def validate(self) -> None:
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "season", "elapsed", "interval", "estimate", "previous"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert all(
+            self[x].dtype == pl.Utf8 for x in ["region", "season"]
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric()
+            for x in ["elapsed", "interval", "estimate", "previous"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class IncidentUptakeProjection(UptakeData):
```

Comment:
Although we're not there yet, I can foresee a situation where "projections" are expected to always include a credible interval about the point estimate, but "data" are not. So a separate class for "projections" that can (in the future) be validated a little differently than "data" seemed wise. If this situation does eventually occur, is there a better way to handle it?

Reply:
The holistic solution would be to have 3 columns that let you specify what kind of estimate is, an estimate ID, and the value

For a point estimate, you have `point_estimate`, no ID, and the number. For CIs, you could have `quantile` and ID `0.025` for the lower CI, eg. For Bayesian posterior sample draws, you could have `posterior_sample` and then IDs that go from 1 to whatever.

In the [forecast hub architecture](https://github.com/reichlab/covid19-forecast-hub/blob/master/data-processed/README.md#type), they have `type` as either `"point"` or `"quantile"`, and then `quantile` is either NA or the quantile. I find this a little restrictive, since it means you need a different table schema for posterior draws.

As a sidebar, the *input* data also has confidence intervals, which we might want for certain kinds of models (that use the input CIs for error on the observation process)